config/script.go

```diff
@@ -0,0 +1,192 @@
+// Copyright 2023 Terramate GmbH
+// SPDX-License-Identifier: MPL-2.0
+
+package config
+
+import (
+	hhcl "github.com/hashicorp/hcl/v2"
+	"github.com/terramate-io/terramate/errors"
+	"github.com/terramate-io/terramate/hcl"
+	"github.com/terramate-io/terramate/hcl/eval"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// Errors for indicating invalid script schema
+const (
+	ErrScriptSchema              errors.Kind = "script config has an invalid schema"
+	ErrScriptInvalidTypeDesc     errors.Kind = "invalid type for script.description"
+	ErrScriptInvalidTypeCommand  errors.Kind = "invalid type for script.command"
+	ErrScriptInvalidTypeCommands errors.Kind = "invalid type for script.commands"
+)
+
+// EvaluatedScriptJob represents an evaluated job block
+type EvaluatedScriptJob struct {
+	Command  []string
+	Commands [][]string
+}
+
+// EvaluatedScript represents an evaluated script block
+type EvaluatedScript struct {
+	Labels      []string
+	Description string
+	Jobs        []EvaluatedScriptJob
+}
+
+// CommandOrCommands is a convenience method for callers who don't specifically
+// care about which command attr was set e.g. command or commands. This method
+// returns a list of commands irrespective of whether they were set through
+// job.command or job.commands
+func (es EvaluatedScriptJob) CommandOrCommands() [][]string {
+	if es.Command != nil {
+		return [][]string{es.Command}
+	}
+	return es.Commands
+}
+
+// EvalScript evaluates a script block using the provided evaluation context
+func EvalScript(evalctx *eval.Context, script hcl.Script) (EvaluatedScript, error) {
+	evaluatedScript := EvaluatedScript{
+		Labels: script.Labels,
+	}
+	errs := errors.L()
+
+	desc, err := evalScriptDesc(evalctx, script.Description.Expr, "script.description")
```

Comment:
we have `evalString()` in this same package.
https://github.com/terramate-io/terramate/blob/main/config/assert.go#L74-L86

Reply:
Yes, I considered using it however I'd like the eval* functions to return specific errors (e.g. not a general error e.g.`ErrSchema`) that are relevant to the attributes e.g. description

Reply:
you can wrap the errors with `errs.AppendWrap()`, or `err.Append(errors.E(OtherErr, err))`.

Reply:
I see, I'll use the `evalString()` function then. Would it be possible to print only the top level error when we later emit this to the user's stdout? 🤔

Reply:
I think the wrapping makes sense (as you did now).

```
terramate schema error: invalid type for script.description
```

Do you think it should be still improved?

Reply:
Ah sorry, I forgot to embed the underlying error, fixed now.
![image](https://github.com/terramate-io/terramate/assets/92498/b671b455-4c58-4c84-9611-4c93a4bedca0)